### ISSUE TITLE
pgrepr: binary interval encoding is wrong

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#4927]
`SELECT INTERVAL '-2147483647 days -2147483647 hours -59 minutes -59.999999 seconds'` over binary returns an incorrect result.

**Comment by jkosh44** [comment]
For reference, the postgres implementation can be found here: https://github.com/postgres/postgres/blob/f48385c1327f1eaef134ad41b382041a79c5200a/src/backend/utils/adt/timestamp.c#L976-L1015

**Comment by jkosh44** [comment]
For reference, the binary encoding of an Interval is the following: [Microseconds (8 bytes), Days (4 bytes), Months (4 bytes)].


The issue is that Postgres will not roll hours into days, even if the amount of hours exceed one day. Materialize on the other hand will roll hours into days if the hours exceed one day. The issue is much easier to see with the following query:

`SELECT INTERVAL '24 hrs';`

Postgres creates the following binary encoding: `[0, 0, 0, 20, 29, 215, 96, 0, 0, 0, 0, 0, 0, 0, 0, 0]`. As you can see, only the microsecond bytes are filled in and the rest are left blank.

Materialize on the other hand creates the following binary encoding: `["0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0"]`. Only the day bytes are filled in with a value of 1, and the rest are left blank.

Interestingly, for both Postgres and Materialize, the following query: `SELECT INTERVAL '1 day'` creates the following binary encoding: `["0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0"]`, which again only has a 1 in the day bytes.

The reason is that postgres doesn't always consider 24 hrs equal to 1 day in the context of intervals, but sometimes it does. This results in some _odd_ behavior in postgres, ex:
```
postgres=# SELECT INTERVAL '24 hrs';
 interval 
----------
 24:00:00
(1 row)

postgres=# SELECT INTERVAL '1 day';
 interval 
----------
 1 day
(1 row)

postgres=# SELECT INTERVAL '1 day' - INTERVAL '24 hours';
    ?column?     
-----------------
 1 day -24:00:00
(1 row)

postgres=# SELECT INTERVAL '1 day' = INTERVAL '24 hours';
 ?column? 
----------
 t
(1 row)

postgres=# SELECT (INTERVAL '1 day' - INTERVAL '24 hours') = INTERVAL '0';
 ?column? 
----------
 t
(1 row)
```

Materialize does always consider 24 hrs equivalent to 1 day:
```
materialize=> SELECT INTERVAL '24 hrs';
 interval 
----------
 1 day
(1 row)

materialize=> SELECT INTERVAL '1 day';
 interval 
----------
 1 day
(1 row)

materialize=> SELECT INTERVAL '1 day' - INTERVAL '24 hours';
 ?column? 
----------
 00:00:00
(1 row)

materialize=> SELECT INTERVAL '1 day' = INTERVAL '24 hours';
 ?column? 
----------
 t
(1 row)

materialize=> SELECT (INTERVAL '1 day' - INTERVAL '24 hours') = INTERVAL '0';
 ?column? 
----------
 t
(1 row)
```

Unfortunately, right now, Materialize needs values that are equal to have the same binary encoding, which fixing this would break. This issue and comment goes into more detail about why that is: https://github.com/MaterializeInc/materialize/pull/3151#issuecomment-634206825. So right now I don't think we can fix this issue.

**Comment by mjibson** [comment]
Does this get fixed if our in memory impl changes to just have micros, days, and months, instead of the current months and nanos?

**Comment by jkosh44** [comment]
Our current implementation only stores nanoseconds and months. If we every want to fix this, then we **need** to change our implementation to store months, days, and micros/nanos (I don't think it matters if we store micros or nanos). Our current implementation has no way to differentiate b/w 24 hours and 1 day because days get flattened down into nanoseconds.

The issue is that if we start to differentiate b/w 24 hours and 1 day, then we run into the issue where `INTERVAL '1 day'` produces a different binary encoding than `INTERVAL '24 hours'` even though they are considered equivalent. My understanding is that currently in Materialize, if two values are considered equivalent, then they MUST have the same binary encodings.

**Comment by mjibson** [comment]
I think the definition of equivalent can change though. 24 hours doesn't have to be equal to 1 day.

**Comment by mjibson** [comment]
I do not know the exact requirements for dataflow beyond that if things are SQL-equivalent, then they must have the same byte encoding in dataflow. Are there any ordering requirements? If not, and it's only the byte equality, then maybe:

For intervals, we implement sql equality as a blind struct comparison, and thus all fields must match, so we continue to have 1 day be not equal to 24 hours (at least until we do #10566). For inequality comparisons, we have standard rules that 24h=1day, 1month=30days=720hours, 1h=60m, 1m=60s, and like convert everything to an i128 micros then compare that. This would allow us to switch to the postgres in memory representation (micros, days, months) and have most stuff work.

**Comment by jkosh44** [comment]
Just repeating here what we discussed earlier for documentation purposes. The issues I can see with this approach is that `ORDER BY`s can become non-deterministic. For example consider the following queries
```
postgres=# CREATE TABLE temp(i INTERVAL);
CREATE TABLE
postgres=# INSERT INTO temp VALUES (INTERVAL '24 hrs'), (INTERVAL '1 day');
INSERT 0 2
postgres=# SELECT DISTINCT * FROM temp ORDER BY i ASC;
```
We could have two potential results to the `SELECT`. The two intervals are not considered equivalent when performing the `DISTINCT` so we'll have two rows, however they are considered equivalent when performing the `ORDER BY`, so any order is correct.

**Comment by jkosh44** [comment]
I don't know a ton about the ordering function, but I think it probably uses <, >, <=, >= behind the scenes. So in order to make it deterministic we would not be able to say that 24 hrs=1 day when comparing two Intervals.

**Comment by mjibson** [comment]
> we would not be able to say that 24 hrs=1 day when comparing two Intervals

yes, i'm totally ok with that because #10566 will allow us to support it in the future.

**Comment by jkosh44** [comment]
Repeating what was said earlier today: It doesn't actually matter if ordering is non-deterministic. So we can do that smart thing and say that 24 hrs == 1 day and 30 days == 1 month, when ordering Intervals.

**Comment by mjibson** [comment]
Did the binary encoding issue here get fixed?

**Comment by jkosh44** [comment]
@mjibson Yep, I added the example in the issue to the binary.pt test as well. I would have thought that my commit message would automatically close this, but I guess not. Closing now.
